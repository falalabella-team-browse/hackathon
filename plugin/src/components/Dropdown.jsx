import React, { useState, useEffect, useRef } from "react";
import styled from "styled-components";
import DownArrow from "./DownArrow";

const Container = styled.div`
  position: relative;
  margin: 12px 0;
  padding: 8px 0;
  border-bottom: 2px solid #dfdfdf;
  max-width: 300px;
  cursor: pointer;
`;

const Selected = styled.span`
  font-weight: bold;
  font-size: 14px;
  padding: 0 12px;
`;

const Options = styled.ul`
  position: absolute;
  top: 40px;
  left: 0;
  right: 0;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  background-color: white;
  list-style-type: none;
`;

const Option = styled.li`
  padding: 12px 12px;
  font-weight: 400;
  font-size: 14px;

  &:hover {
    background-color: #dfdfdf;
  }
`;

const Icon = styled.div`
  position: absolute;
  top: 10px;
  right: 15px;
`;

const Dropdown = ({ options = [], selected, onChange }) => {
  const [focus, setFocus] = useState(false);
  const wrapperRef = useRef();

  const selectedOption =
    options.find((opt) => opt.value === selected) || options[0];

  const handleOnClick = () => {
    setFocus(!focus);
  };

  const handleClickOutside = (e) => {
    if (wrapperRef && !wrapperRef.current.contains(e.target)) {
      setFocus(false);
    }
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  });

  const handleOptionSelected = (value) => () => {
    onChange(value);
  };

  return (
    <Container onClick={handleOnClick} ref={(r) => (wrapperRef.current = r)}>
      <Selected>{selectedOption.label}</Selected>
      <Icon>
        <DownArrow />
      </Icon>
      {focus && (
        <Options>
          {options.map(({ label, value }) => (
            <Option onClick={handleOptionSelected(value)} key={value}>
              {label}
            </Option>
          ))}
        </Options>
      )}
    </Container>
  );
};

export default Dropdown;
