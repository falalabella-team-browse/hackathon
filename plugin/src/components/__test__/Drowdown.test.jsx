import React from "react";
import { mount, shallow } from "enzyme";
import Dropdown from "../Dropdown";
import { act } from "react-dom/test-utils";

const OPTION = [
  { label: "Option 1", value: "v1" },
  { label: "Option 2", value: "v2" },
];

describe("Dropdown", () => {
  it("should render properly without any props", () => {
    const fn = jest.fn();
    const wrapper = mount(<Dropdown onChange={fn} />);

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("should change value on click", () => {
    const fn = jest.fn();
    const wrapper = mount(<Dropdown options={OPTION} onChange={fn} />);

    wrapper.find("div").first().simulate("click");
    wrapper.find("li").first().simulate("click");

    expect(fn).toBeCalled;
    wrapper.unmount();
  });

  it("should close dropdown when focus is lost", () => {
    const mock = {};
    const oal = document.addEventListener;

    document.addEventListener = (type, cb) => {
      mock[type] = cb;
    };

    const fn = jest.fn();
    const wrapper = mount(
      <section>
        <a>link</a>
        <Dropdown options={OPTION} onChange={fn} />
      </section>
    );

    expect(mock.mousedown).toBeDefined();

    wrapper.find("div").first().simulate("click");
    expect(wrapper.find("li").length).toBe(2);

    act(() => {
      mock.mousedown({
        target: document.createElement("p"),
      });
    });

    wrapper.update();

    expect(wrapper.find("li").length).toBe(0);

    document.addEventListener = oal;
  });
});
