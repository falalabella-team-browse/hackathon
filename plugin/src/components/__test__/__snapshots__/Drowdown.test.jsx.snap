// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render properly without any props 1`] = `
"<Dropdown onChange={[Function: mockConstructor]}>
  <Dropdown__Container onClick={[Function: handleOnClick]}>
    <div onClick={[Function: handleOnClick]} className=\\"Dropdown__Container-rnr-__hauupl-0 fJWjsI\\">
      <Dropdown__Selected>
        <span className=\\"Dropdown__Selected-rnr-__hauupl-1 kSECSH\\" />
      </Dropdown__Selected>
      <Dropdown__Icon>
        <div className=\\"Dropdown__Icon-rnr-__hauupl-4 fVRfmr\\">
          <DownArrow>
            <svg version=\\"1.1\\" id=\\"Layer_1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0px\\" y=\\"0px\\" viewBox=\\"0 0 512.011 512.011\\" height=\\"15px\\" width=\\"15px\\" style={{...}}>
              <g>
                <g>
                  <path fill=\\"#000\\" d=\\"M505.755,123.592c-8.341-8.341-21.824-8.341-30.165,0L256.005,343.176L36.421,123.592c-8.341-8.341-21.824-8.341-30.165,0 s-8.341,21.824,0,30.165l234.667,234.667c4.16,4.16,9.621,6.251,15.083,6.251c5.462,0,10.923-2.091,15.083-6.251l234.667-234.667 C514.096,145.416,514.096,131.933,505.755,123.592z\\" />
                </g>
              </g>
            </svg>
          </DownArrow>
        </div>
      </Dropdown__Icon>
    </div>
  </Dropdown__Container>
</Dropdown>"
`;
